Optional stress dependencies & don't populate CORE_ROOT twice

trylek opened the pull request:
Bruce noticed that we're downloading and installing stress
runtime dependencies several times during build. I have added a new
option to suppress this behavior in build-test and I fixed the
invocations of build-test based on Bruce's analysis. I have also
patched the build scripts to avoid populating CORE_ROOT in
"copynativeonly" mode.

Thanks

Tomas

Fixes: #36797

/cc: @dotnet/runtime-infrastructure

trylek committed:
Optional stress dependencies & don't populate CORE_ROOT twice

Bruce noticed that we're downloading and installing stress
runtime dependencies several times during build. I have added a new
option to suppress this behavior in build-test and I fixed the
invocations of build-test based on Bruce's analysis. I have also
patched the build scripts to avoid populating CORE_ROOT in
"copynativeonly" mode.

Thanks

Tomas

Fixes: #36797

trylek committed:
Indentation

trylek committed:
Enable package restoration during copynativeonly

BruceForstall commented on src/coreclr/build-test.cmd:
```diff
@@ -98,13 +99,16 @@ if /i "%1" == "checked"               (set __BuildType=Checked&set processedArgs
 
 if /i "%1" == "ci"                    (set __ArcadeScriptArgs="-ci"&set __ErrMsgPrefix=##vso[task.logissue type=error]&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
 
+if /i "%1" == "skipstressdependencies" (set __SkipStressDependencies=1&shift&goto Arg_Loop)
```
Don't you also need the `&set processedArgs=!processedArgs! %1` magic?

BruceForstall commented on src/coreclr/build-test.cmd:
```diff
@@ -280,8 +287,6 @@ if errorlevel 1 (
     exit /b 1
 )
 
-:skipnative
```
Why is this removed?

trylek replied:
I believe it's not used anywhere. Perhaps I'm just blind, are you aware of any goto or call statement referring to this label?

BruceForstall replied:
Just above at the top of this "section": "if defined __SkipNative goto skipnative"

trylek replied:
Hmm, my bad, thanks for pointing that out, for some reason I was under the impression I saw two instances of the label in the file. I probably need to buy myself stronger glasses :-). Will put back...

BruceForstall reviewed (changes requested):
Looks like a problem with your change is that you changed the "copynativeonly" phase to not generate the layouts (good) (and maybe "copynativeonly" should imply "skipstressdependencies" without having to pass that argument as well?)

Then, the Core_Root needs to be created with the "buildtestwrappersonly" step. On Windows, build-test.cmd does this. But on Linux, build-test.sh does not (which actually makes more sense, given the argument is "...only").

BruceForstall commented:
Note also that in the R2R case, the "Crossgen framework assemblies" step (which follows the steps you modified) _also_ downloads the stress dependencies *and* generates Core_Root (again) (at least on Windows).

trylek committed:
Cleanly separate CORE_ROOT generation and framework crossgenning

I have added a new option for only crossgenning the framework and
I fixed the build scripts to use it. I have added a separate
pipeline step for populating CORE_ROOT so that it no longer needs
to party on other build steps. My longer term thinking is that
I would love to move most of this logic to MSBuild and then the
individual steps will just correspond to targets or properties
controlling the individual "test build pipeline" steps.

Thanks

Tomas

BruceForstall reviewed (approved):
LGTM

trylek commented:
The AllConfiguration libraries build error looks very similar to

https://github.com/dotnet/runtime/pull/36285

I'm following up to figure out whether it regressed again.
